File: Algorithms/54. Spiral Matrix/javascript/solution.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function (matrix) {
    const result = [];
    const rows = matrix.length, columns = matrix[0].length;
    let up = 0, left = 0;
    let right = columns - 1;
    let down = rows - 1;

    while (result.length < rows * columns) {
        for (let col = left; col <= right; col++) {
            result.push(matrix[up][col]);
        }

        for (let row = up + 1; row <= down; row++) {
            result.push(matrix[row][right]);
        }

        if (up !== down) {
            for (let col = right - 1;  left - 1 < col; col--) {
                result.push(matrix[down][col]);
            }
        }

        if (left !== right) {
            for (let row = down - 1; up < row; row--) {
                result.push(matrix[row][left]);
            }
        }

        left++;
        right--;
        up++;
        down--;
    }

    return result;
};

File: Algorithms/54. Spiral Matrix/kotlin/solution.kt
class Solution {
    fun spiralOrder(matrix: Array<IntArray>): List<Int> {
        val order = mutableListOf<Int>()
        if (matrix.isEmpty()) {
            return order
        }

        var startRow = 0
        var endRow = matrix.size - 1
        var startColumn = 0
        var endColumn = matrix[0].size - 1

        while (startRow <= endRow && startColumn <= endColumn) {
            // Right
            for (column in startColumn..endColumn) {
                order.add(matrix[startRow][column])
            }

            // Down
            for (row in startRow + 1..endRow) {
                order.add(matrix[row][endColumn])
            }

            // Left
            for (column in endColumn - 1 downTo startColumn) {
                if (startRow == endRow) break
                order.add(matrix[endRow][column])
            }

            // Up
            for (row in endRow - 1 downTo startRow + 1) {
                if (startColumn == endColumn) break
                order.add(matrix[row][startColumn])
            }

            startRow++
            endRow--
            startColumn++
            endColumn--
        }

        return order
    }
}
